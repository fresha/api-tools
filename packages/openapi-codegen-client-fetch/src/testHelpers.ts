// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
