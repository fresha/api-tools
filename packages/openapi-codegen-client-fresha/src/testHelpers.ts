// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { getAPIName, getRootUrlOrThrow } from '@fresha/openapi-codegen-utils';

import type { Context } from './context';
import type { OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel, rootDir = '/'): Context => {
  const base = createTSProjectTestContext(openapi, rootDir);

  let apiRootUrl: string;
  try {
    apiRootUrl = getRootUrlOrThrow(openapi);
  } catch (err) {
    apiRootUrl = 'API_URL';
  }

  return {
    ...base,
    apiName: getAPIName(openapi),
    apiRootUrl,
  };
};
