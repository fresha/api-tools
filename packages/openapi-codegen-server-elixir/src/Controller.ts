import console from 'console';

import type { Action } from './Action';
import type { Generator } from './Generator';
import type { PathItemModel } from '@fresha/openapi-model/build/3.0.3';

export class Controller {
  static makeModuleName(pathUrl: string): string {
    return '' ?? pathUrl;
  }

  static makeFileName(): string {
    return '';
  }

  private readonly generator: Generator;
  private readonly actions: Action[];

  constructor(generator: Generator) {
    this.generator = generator;
    this.actions = [];
  }

  generateCode(): void {
    console.log(!!this.generator);
    for (const action of this.actions) {
      action.generateCode();
    }
  }

  // eslint-disable-next-line class-methods-use-this
  processPathItem(pathUrl: string, pathItem: PathItemModel): void {
    console.log(pathUrl, pathItem);
  }
}
