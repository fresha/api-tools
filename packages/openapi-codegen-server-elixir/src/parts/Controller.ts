import assert from 'assert';

import { Action } from './Action';

import type { Context } from './types';
import type { SourceFile } from '@fresha/ex-morph';
import type {
  OperationModel,
  PathItemModel,
  PathItemOperationKey,
} from '@fresha/openapi-model/build/3.0.3';

const actionNameFromOperationKey = (
  operationKey: PathItemOperationKey,
  operation: OperationModel,
): string => {
  switch (operationKey) {
    case 'get':
      return operation.getParameter('id', 'path') ? 'show' : 'index';
    case 'post':
      return 'create';
    case 'put':
    case 'patch':
      return 'update';
    case 'delete':
      return 'delete';
    default:
      assert.fail(`Unsupported method ${operationKey}`);
  }
  return '';
};

export class Controller {
  readonly context: Context;
  readonly moduleName: string;
  readonly sourceFile: SourceFile;
  protected readonly actions: Map<string, Action>;

  constructor(context: Context, moduleName: string) {
    this.context = context;
    this.moduleName = moduleName;
    this.sourceFile = this.context.project.createControllerFile(moduleName);
    this.actions = new Map<string, Action>();
  }

  actionEntries(): IterableIterator<[string, Action]> {
    return this.actions.entries();
  }

  collectData(pathItem: PathItemModel): void {
    this.context.logger.info(`Collecting data for the controller: "${this.moduleName}"`);

    for (const [operationKey, operation] of pathItem.operations()) {
      const actionName = actionNameFromOperationKey(operationKey, operation);
      assert(!this.actions.has(actionName));
      const action = new Action(this.context, this.sourceFile, actionName, operationKey, operation);
      this.actions.set(actionName, action);
    }
  }

  generateCode(): void {
    this.context.logger.info(`Generating code of the controller: "${this.moduleName}"`);

    this.sourceFile.writeDefmodule(this.moduleName, () => {
      this.sourceFile.writeUse(this.context.project.getAppModuleName(), ':controller');
      this.sourceFile.newLine();
      this.sourceFile.writeFunctionCall(
        'action_fallback',
        this.context.project.getControllerModuleName('fallback'),
      );
      this.sourceFile.newLine();
      this.sourceFile.writeLine('# add aliases here');

      for (const action of this.actions.values()) {
        action.generateCode();
      }

      for (const action of this.actions.values()) {
        if (action.needsDateTimeParser) {
          this.generateNaiveDateTimeParsers();
          break;
        }
      }
    });
  }

  protected generateNaiveDateTimeParsers(): void {
    this.sourceFile.newLine();
    this.sourceFile.writeLine('defp naive_date_time(nil), do: {:error, :invalid_date_time}');
    this.sourceFile.newLine();
    this.sourceFile.writeLine('defp naive_date_time(input) do');
    this.sourceFile.writeIndented(() => {
      this.sourceFile.writeLine('case NaiveDateTime.from_iso8601(input) do');
      this.sourceFile.writeIndentedLines(
        '{:ok, date_time} -> {:ok, date_time}',
        '{:error, _} -> {:error, :invalid_date_time}',
      );
      this.sourceFile.writeLine('end');
    });
    this.sourceFile.writeLine('end');
  }
}
