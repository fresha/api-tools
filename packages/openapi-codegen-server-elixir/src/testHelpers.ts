import { Project } from '@fresha/code-morph-ex';
import { createRegistry } from '@fresha/json-api-model';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTestContext as createBaseTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory } from '@fresha/openapi-model/build/3.0.3';

import { Generator } from './parts';

import type { Context } from './context';

export const createTestContext = (phoenixApp: string, rootDir = '/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createBaseTestContext(openapi, rootDir);

  const project = new Project({
    rootDir,
    phoenixApp,
    useInMemoryFileSystem: true,
  });

  return {
    ...base,
    testObjectFactoryModuleName: `${project.getAppModuleName()}.Factory`,
    openapi,
    project,
    registry: createRegistry(),
  };
};

export const createGenerator = (phoenixApp = 'api_tools_web'): Generator => {
  const context = createTestContext(phoenixApp, '/');
  const generator = new Generator(context);
  return generator;
};
