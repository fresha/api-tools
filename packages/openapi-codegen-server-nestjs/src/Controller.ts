import path from 'path';

import { Action } from './Action';
import {
  addCommonNestImports,
  addDecorator,
  commonStringPrefix,
  kebabCase,
  significantNameParts,
  startCase,
} from './utils';

import type { Generator } from './Generator';
import type { Logger } from './utils/logging';
import type { PathItemModel } from '@fresha/openapi-model/build/3.0.3';
import type { SourceFile } from 'ts-morph';

/**
 * Generates code for a single NestJS controller.
 */
export class Controller {
  /**
   * Given OpenAPI path item URL, returns controller class name.
   */
  static makeClassName(pathUrl: string): string {
    const nameParts = significantNameParts(pathUrl);
    return startCase(nameParts.concat('Controller').join('-')).replace(/\s+/g, '');
  }

  /**
   * Given an OpenAPI path item URL, returns controller file name.
   */
  static makeFileName(pathUrl: string): string {
    const nameParts = significantNameParts(pathUrl);
    if (nameParts.length) {
      return `${kebabCase(nameParts.join('-'))}.controller.ts`;
    }
    return 'controller.ts';
  }

  readonly generator: Generator;
  readonly className: string;
  readonly outputPath: string;
  protected readonly tsSourceFile: SourceFile;
  protected readonly actions: Action[];
  protected urlPrefix: string | null;
  protected readonly logger: Logger;

  constructor(generator: Generator, outputPath: string, className: string, logger: Logger) {
    this.generator = generator;
    this.outputPath = outputPath;
    this.className = className;
    this.tsSourceFile = this.generator.tsProject.createSourceFile(this.outputPath, '', {
      overwrite: true,
    });
    this.actions = [];
    this.urlPrefix = null;
    this.logger = logger;
  }

  relativeModulePath(filePath: string): string {
    return `./${path.relative(
      path.dirname(this.outputPath),
      path.join(path.dirname(filePath), path.basename(filePath, '.ts')),
    )}`;
  }

  processPathItem(pathUrl: string, pathItem: PathItemModel): void {
    for (const [methodName, operation] of pathItem.operations()) {
      this.actions.push(new Action(this, pathUrl, methodName, operation, this.logger));
    }
    this.urlPrefix = null;
  }

  getUrlPrefix(): string {
    if (this.urlPrefix == null) {
      this.urlPrefix = commonStringPrefix(this.actions.map(action => action.pathUrl));
    }
    return this.urlPrefix;
  }

  generateCode(): void {
    this.logger.info(`Generating controller code for ${this.outputPath}`);

    addCommonNestImports(this.tsSourceFile, 'Controller');

    const classDecl = this.tsSourceFile.addClass({
      name: this.className,
      isExported: true,
    });

    addDecorator(classDecl, 'Controller', this.getUrlPrefix());

    for (const action of this.actions) {
      action.generateCode(classDecl);
    }
  }
}
