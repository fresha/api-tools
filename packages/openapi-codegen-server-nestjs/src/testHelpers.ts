import { Nullable } from '@fresha/api-tools-core';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory, SchemaModel } from '@fresha/openapi-model/build/3.0.3';

import { Context } from './context';
import { Generator } from './Generator';

export const createTestContext = (nestApp = 'web', outputPath = '/tmp/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createTSProjectTestContext(openapi, outputPath);

  base.project.createSourceFile(
    `${outputPath}/${nestApp}.module.ts`,
    `import { Module } from '@nestjs/common';
    import { AppController } from './app.controller';
    import { AppService } from './app.service';

    @Module({
      imports: [],
      controllers: [AppController],
      providers: [AppService],
    })
    export class AppModule {}`,
  );

  return {
    ...base,
    nestApp,
    addDTO: jest.fn(),
  };
};

export const createGenerator = (nestApp = 'app', rootDir = '/'): Generator => {
  const context = createTestContext(nestApp, rootDir);

  let generator: Generator;
  // eslint-disable-next-line prefer-const
  generator = new Generator({
    ...context,
    addDTO: (name: string, schema: Nullable<SchemaModel>) => generator.addDTO(name, schema),
  });

  return generator;
};
