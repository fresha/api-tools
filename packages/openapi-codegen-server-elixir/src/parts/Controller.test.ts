import {
  addResourceRelationship,
  MEDIA_TYPE_JSON_API,
  RelationshipCardinality,
  setDataDocumentSchema,
} from '@fresha/openapi-codegen-utils';

import { createTestContext } from '../testHelpers';

import { Controller } from './Controller';

test('parameter-less action, minimal implementation', () => {
  const context = createTestContext('awesome_web');

  const pathItem = context.openapi.setPathItem('/employees/:id/tasks');
  const operation = pathItem.addOperation('post');
  operation.operationId = 'createEmployee';

  const controller = new Controller(context, '/pracownicy', 'AwesoneWeb.EmployeesController');
  controller.collectData(pathItem);
  controller.generateCode();

  expect(controller.sourceFile.getText()).toMatchSnapshot();
});

test('ID parameter leads to generating parse_XXXX_param function, as well as error handling', () => {
  const context = createTestContext('awesome_web');

  const pathItem = context.openapi.setPathItem('/employees/:id/tasks');
  const operation = pathItem.addOperation('get');
  operation.operationId = 'readEmployee';
  operation.addParameter('id', 'path').setSchema('string');

  const controller = new Controller(context, 'employees', 'EmployeeController');
  controller.collectData(pathItem);
  controller.generateCode();

  expect(controller.sourceFile.getText()).toMatchSnapshot();
});

test('request body leads to generating parse_XXXX_conn function, as well as error handling', () => {
  const context = createTestContext('awesome_web');

  const pathItem = context.openapi.setPathItem('/profile');
  const requestBodySchema = pathItem
    .addOperation('put')
    .setRequestBody()
    .setMediaType(MEDIA_TYPE_JSON_API)
    .setSchema('object');
  setDataDocumentSchema(requestBodySchema, 'profile-settings');
  const attributesSchema = requestBodySchema
    .getPropertyOrThrow('data')
    .getPropertyDeepOrThrow('attributes');
  attributesSchema.setProperties({
    name: { type: 'string', required: true },
    age: { type: 'integer', required: true, minimum: 18, maximum: 200 },
    weight: 'float',
    birthDate: 'date',
    score: 'number',
    gender: { type: 'string', required: true, enum: ['male', 'female', 'other'] },
    num1: { type: 'number', minimum: 10, exclusiveMinimum: true },
    num2: { type: 'number', minimum: 10, exclusiveMinimum: false },
    num3: { type: 'number', maximum: 20, exclusiveMaximum: true },
    num4: { type: 'number', maximum: 20, exclusiveMaximum: false },
    int1: { type: 'integer', minimum: 10, exclusiveMinimum: true },
    int2: { type: 'integer', minimum: 10, exclusiveMinimum: false },
    int3: { type: 'integer', maximum: 20, exclusiveMaximum: true },
    int4: { type: 'integer', maximum: 20, exclusiveMaximum: false },
  });

  const resourceSchema = requestBodySchema.getPropertyOrThrow('data');
  addResourceRelationship(resourceSchema, 'location', 'locations');
  addResourceRelationship(
    resourceSchema,
    'employee',
    'employees',
    RelationshipCardinality.One,
    false,
  );

  const controller = new Controller(context, '/profile', 'AwesomeWeb.ProfileController');
  controller.collectData(pathItem);
  controller.generateCode();

  // TODO test
  // :decimal
  // :email
  // trim options for :string
  expect(controller.sourceFile.getText()).toMatchSnapshot();
});
