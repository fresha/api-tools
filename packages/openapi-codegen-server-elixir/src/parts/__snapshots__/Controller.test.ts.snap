// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ID parameter leads to generating parse_XXXX_param function, as well as error handling 1`] = `
"defmodule EmployeeController do
  @moduledoc false

  use AwesomeWeb, :controller

  action_fallback(AwesomeWeb.FallbackController)

  # add aliases here

  def show(conn, params) do
    with {:ok, id} <- parse_show_params(params) do
      # TODO this is the part you need to implement by yourself
      # TODO evaluate extra arguments, then pass them to render()
      render(conn)
    else
      {:error, :invalid_parameters, params} ->
        {:error, :invalid_parameters, params}

      {:error, :invalid_pointers, pointers} ->
        {:error, :invalid_parameters, pointers}

      {:error, :not_found} ->
        {:error, :not_found}
    end
  end

  defp parse_show_params(params) do
    flat_parse(
      params,
      id: [:id, :required],
    )
  end
end"
`;

exports[`parameter-less action, minimal implementation 1`] = `
"defmodule AwesoneWeb.EmployeesController do
  @moduledoc false

  use AwesomeWeb, :controller

  action_fallback(AwesomeWeb.FallbackController)

  # add aliases here

  def create(conn, _params) do
    # TODO this is the part you need to implement by yourself
    # TODO evaluate extra arguments, then pass them to render()
    render(conn)
  end
end"
`;

exports[`request body leads to generating parse_XXXX_conn function, as well as error handling 1`] = `
"defmodule AwesomeWeb.ProfileController do
  @moduledoc false

  use AwesomeWeb, :controller

  action_fallback(AwesomeWeb.FallbackController)

  # add aliases here

  def update(conn, _params) do
    with {:ok, parsed_opts} <- parse_update_conn(conn) do
      # TODO this is the part you need to implement by yourself
      # TODO evaluate extra arguments, then pass them to render()
      render(conn)
    else
      {:error, :invalid_parameters, params} ->
        {:error, :invalid_parameters, params}

      {:error, :invalid_pointers, pointers} ->
        {:error, :invalid_parameters, pointers}

      {:error, :not_found} ->
        {:error, :not_found}
    end
  end

  defp parse_update_conn(conn) do
    parse(
      conn.assigns[:doc],
      id: [:id, :required],
      attributes: %{
        name: [:string, :required],
        age: [{:integer, min: 18, max: 200}, :required],
        weight: :float,
        birth_date: :date,
        score: :float,
        gender: [:string, :required, {:contain, ~w{male female other}}],
        num1: {:float, min: 10},
        num2: {:float, min: 10},
        num3: {:float, max: 20},
        num4: {:float, max: 20},
        int1: {:integer, min: 11},
        int2: {:integer, min: 10},
        int3: {:integer, max: 19},
        int4: {:integer, max: 20},
      },
      relationships: %{
        location: [:resource_id, :required],
        employee: :resource_id,
      },
    )
  end
end"
`;
